CRVResponse/src/MakeCrvRecoPulses.cc

```diff
@@ -30,75 +36,106 @@ void MakeCrvRecoPulses::SetWaveform(const std::vector<unsigned int> &waveform, u
+    }
+    if(waveform[bin-1]==waveform[bin]) //potentially a peak where consecutive ADC values are equal
+    {
+      peakEndBin=bin;
+    }
+    if(waveform[bin-1]>waveform[bin]) //falling edge
+    {
+      if(peakStartBin>0)  //found a peak
+      {
+        peaks.emplace_back(peakStartBin,peakEndBin);
+        peakStartBin=0;  //so that the loop as to wait for the next rising edge
+      }
+    }
   }
 
   for(size_t i=0; i<peaks.size(); i++)
   {
   //select a range of up to 4 points before and after the maximum point
   //-find up to 5 points before and after the maximum point for which the waveform is stricly decreasing
   //-remove 1 point on each side. this removes potentially "bad points" belonging to a second pulse (i.e. in double pulses)
-    int maxBin = peaks[i].first;
-    if(waveform[maxBin]-pedestal<5) continue; //FIXME: need a better way to identify these fake pulse which are caused by electronic noise
+    peakStartBin=peaks[i].first;
+    peakEndBin=peaks[i].second;
+    if(waveform[peakStartBin]-pedestal<5) continue; //FIXME: need a better way to identify these fake pulse which are caused by electronic noise
 
-    int startBin=maxBin;
-    int endBin=maxBin;
-    for(int bin=maxBin-1; bin>=0 && bin>=maxBin-5; bin--)
+    int startBin=peakStartBin;
+    int endBin=peakEndBin;
+    for(int bin=peakStartBin-1; bin>=0 && bin>=peakStartBin-5; bin--)
     {
       if(waveform[bin]<=waveform[bin+1]) startBin=bin;
       else break;
     }
-    for(int bin=maxBin+1; bin<nBins && bin<=maxBin+5; bin++)
+    for(int bin=peakEndBin+1; bin<nBins && bin<=peakEndBin+5; bin++)
     {
       if(waveform[bin]<=waveform[bin-1]) endBin=bin;
       else break;
     }
-    if(maxBin-startBin>1) startBin++;
-    if(endBin-maxBin>1) endBin--;
+    if(peakStartBin-startBin>1) startBin++;
+    if(endBin-peakEndBin>1) endBin--;
 
     double t1=(startTDC+startBin)*digitizationPeriod;
     double t2=(startTDC+endBin)*digitizationPeriod;
+    double peakTime=(startTDC+0.5*(peakStartBin+peakEndBin))*digitizationPeriod;
 
     //fill the graph
-    TGraph g;
+    std::vector<double> t,v;
     for(int bin=startBin; bin<=endBin; bin++) 
     {
-      double t=(startTDC+bin)*digitizationPeriod;
-      double v=waveform[bin]-pedestal;
-      g.SetPoint(g.GetN(), t, v);
+      t.emplace_back((startTDC+bin)*digitizationPeriod);
+      v.emplace_back(waveform[bin]-pedestal);
     }
+    TGraph g(t.size(), t.data(), v.data());
 
     //set the fit function
-    TF1 f("peakfitter","[0]*(TMath::Exp(-(x-[1])/[2]-TMath::Exp(-(x-[1])/[2])))");
-    f.SetParameter(0, (waveform[maxBin]-pedestal)*TMath::E());
-    f.SetParameter(1, (startTDC+maxBin)*digitizationPeriod);
-    f.SetParameter(2, darkNoise?12.6:19.0);
-    if(peaks[i].second) f.SetParameter(1, (startTDC+maxBin+0.5)*digitizationPeriod);
+    _f.SetParameter(0, (waveform[peakStartBin]-pedestal)*TMath::E());
+    _f.SetParameter(1, peakTime);
+    _f.SetParameter(2, darkNoise?12.6:19.0);
```

Comment:
Make these magic numbers symbolic constants.  

It would be better to get them from conditions ( or fcl depending exactly on their meaning, but probably conditions.)  You can do this change later.

Reply:
Done